src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala

```diff
@@ -84,28 +84,41 @@ object InstanceUpdater extends StrictLogging {
 
   private[marathon] def launchOnReservation(instance: Instance, op: LaunchOnReservation): InstanceUpdateEffect = {
     if (instance.isReserved) {
-      require(instance.tasksMap.size == 1, "Residency is not yet implemented for task groups")
+      val currentTasks = instance.tasksMap
+      val taskEffects = currentTasks.map {
+        case (taskId, task) =>
+          val newTaskId = op.oldToNewTaskIds.getOrElse(
+            taskId,
+            throw new IllegalStateException("failed to retrieve a new task ID"))
+          val status = op.statuses.getOrElse(
+            newTaskId,
+            throw new IllegalStateException("failed to retrieve a task status"))
+          task.update(TaskUpdateOperation.LaunchOnReservation(newTaskId, op.runSpecVersion, status))
+      }
 
-      // TODO(PODS): make this work for taskGroups
-      val currentTask: Task = instance.appTask
-      val taskEffect = currentTask.update(TaskUpdateOperation.LaunchOnReservation(op.newTaskId, op.runSpecVersion, op.status))
-      taskEffect match {
-        case TaskUpdateEffect.Update(updatedTask) =>
-          val updated = instance.copy(
-            state = instance.state.copy(
-              condition = Condition.Staging,
-              since = op.timestamp
-            ),
-            tasksMap = Map(updatedTask.taskId -> updatedTask),
-            runSpecVersion = op.runSpecVersion,
-            // The AgentInfo might have changed if the agent re-registered with a new ID after a reboot
-            agentInfo = op.agentInfo
-          )
-          val events = eventsGenerator.events(updated, task = None, op.timestamp, previousCondition = Some(instance.state.condition))
-          InstanceUpdateEffect.Update(updated, oldState = Some(instance), events)
+      val nonUpdates = taskEffects.filter {
+        case _: TaskUpdateEffect.Update => false
+        case _ => true
+      }
```

Comment:
consider using collect?

```
taskEffects.collect { case update: TaskUpdateEffect.Update => update }
```

Reply:
Why? It would collect update ops, whereas here it has to collect non-update ops. Both `nonUpdates` and `allUpdates` are needed.